src/main/groovy/org/transmartproject/batch/clinical/facts/ClinicalFactsRowSetFactory.groovy

```diff
@@ -103,17 +121,56 @@ class ClinicalFactsRowSetFactory {
             if (conceptType == ConceptType.NUMERICAL && !curValIsNumerical) {
                 throw new IllegalArgumentException("Variable $var inferred or specified " +
                         "numerical, but got value '$value'. Patient id: " +
-                        "${result.patient.id}.")
+                        "${variables.getPatientId(row)}.")
             }
 
         }
 
         // we need a subnode if the variable is categorical
         if (conceptType == ConceptType.CATEGORICAL) {
-            concept = tree.getOrGenerate(var.conceptPath + value, ConceptType.CATEGORICAL)
+            concept = tree.getOrGenerate(conceptPath + value, ConceptType.CATEGORICAL)
+        }
+
+        concept
+    }
+
+    private ConceptPath getOrGenerateConceptPath(ClinicalDataFileVariables variables,
+                                       ClinicalVariable var,
+                                       ClinicalDataRow row) {
+        if (var.conceptPath) {
+            return var.conceptPath
         }
 
-        result.addValue concept, getXtrialNodeFor(concept), value
+        assert var.dataLabel == ClinicalVariable.TEMPLATE
+
+        def relConceptPath = ClinicalVariable.toPath(var.categoryCode)
+
+        if (relConceptPath.contains(ClinicalVariable.SITE_ID_PLACEHOLDER)) {
+            def replaceValue = variables.getSiteId(row) ?: ''
+            relConceptPath = relConceptPath.replace(ClinicalVariable.SITE_ID_PLACEHOLDER, replaceValue)
+        }
+
+        if (relConceptPath.contains(ClinicalVariable.VISIT_NAME_PLACEHOLDER)) {
+            def replaceValue = variables.getVisitName(row) ?: ''
+            relConceptPath = relConceptPath.replace(ClinicalVariable.VISIT_NAME_PLACEHOLDER, replaceValue)
+        }
+
+        ClinicalVariable referencedDataLabelVariable = null
+        if (var.dataLabelSource) {
+            referencedDataLabelVariable = variables.dataLabelsColumnNumberIndex.get(var.dataLabelSource)
+        } else {
+            referencedDataLabelVariable = variables.dataLabelsColumnNumberIndex.values().iterator().next()
```

Comment:
@cataphract There must be single data label variable at the point when we take first element from `referencedDataLabelVariable `. Otherwise the workflow would complain on column mapping file here https://github.com/thehyve/transmart-batch/blob/dynamic-concepts/src/main/groovy/org/transmartproject/batch/clinical/facts/ClinicalDataFileVariables.groovy#L62